### PR TITLE
[Feature] Add per-request attention capture to the OpenAI-compatible API

### LINKED ISSUE #11365: [Feature]: Add support for attention score output
## Problem
vLLM currently doesn't provide access to attention scores during inference, which are essential for model analysis and interpretability research. #11862 

## Feature Request
Add the ability to retrieve attention scores during model inference, similar to HuggingFace's output_attentions=True parameter.

## Motivation
Need to analyze token-level relationships in model outputs
Required for building visualization tools and debugging model behavior
Critical for research into attention mechanisms

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by Parkprogrammer**
Closes #11365

## Motivation

Interpretability researchers and multimodal debugging workflows need access to raw attention scores at inference time without patching model code or writing custom inference loops.

This PR exposes per-request attention instrumentation through the existing OpenAI-compatible API. When not requested, there is zero additional compute cost — no forward-pass modification, no graph break, no torch.compile impact.

---

## What This PR Does

Adds an opt-in, per-request attention capture mechanism to vLLM's OpenAI API server. When the server is started with instrumentation enabled and a client explicitly requests capture, Q×Kᵀ softmax attention scores are computed
post-forward, serialised via shared memory, and attached to the chat completion response.

---

## New Server Flags

```bash
--enable-attention-instrumentation        # Enable feature (disabled by default)
--attention-instrumentation-layers all   # Default capture layer set (overridable per-request)
```

---

## New Request Fields (Chat Completions)

| Field | Type | Description |
|-------|------|-------------|
| `attn_capture` | `int` (0/1) | Enable capture for this request |
| `attn_capture_layers` | `str` | Comma-separated layer indices or `"all"` |

---

## New Response Field

`attn_capture_data` — list of per-layer objects:

```json
[
  {
    "layer_idx": 8,
    "data": "<base64-gzipped tensor>",
    "shape": [T, H, T],
    "dtype": "float16",
    "token_meta": {
      "token_idx_basis": 0,
      "prompt_len": 42,
      "total_len": 50,
      "vision_ranges": [[0, 35]]
    }
  }
]
```

Shape is `[T, H, T]` where `T` = `prompt_len + generated_len - 1` (all tokens visible to the attention kernel — the final sampled token has no subsequent Q step). `H` = query heads.

`token_meta.vision_ranges` maps image token spans for cross-modal analysis.

---

## Design Notes

**Why post-forward computation?**
- Scores are computed after the forward pass completes, leaving model kernels and the attention backend entirely unmodified. No graph break or `torch.compile` side effect is introduced.

**Why Q buffering?**
- Q tensors are captured per-slot during the forward pass and stored in a per-worker buffer. After the request finishes, <br/>K is read from the KV cache and Q×Kᵀ is computed once. This avoids any impact on the hot path.

**Prefix caching interaction**
- When prefix caching is active, prompt tokens are cache-hit and their Q vectors are not recomputed. A persistent `q_cache` (FIFO-capped at 32 768 slots, ~128 MB at float16/128d/16h) stores Q vectors across requests so prefix-cached tokens are still included in the captured score matrix.

**Why shared memory?**
- Attention tensors can be large. Shared memory avoids copying them through the engine's RPC transport layer. <br/>Each snapshot is keyed by `request_id` and cleaned up immediately after the output processor reads it.

**Concurrency and isolation**
- Each request uses its own shared-memory segment named by `request_id`, so concurrent requests are fully isolated. Segments are unlinked after reading; a 30-second read timeout is used to handle slow capture paths gracefully. <br/>If the engine crashes before writing, the segment simply times out and `attn_capture_data` is absent from the response (no server crash).

**Streaming compatibility**
- Attention scores are only available in non-streaming responses (`stream=False`). Streaming responses do not include `attn_capture_data` because the snapshot is only available after the full sequence is generated.

---

## Memory Impact

Per captured request (non-streaming only):

| T (seq len) | H (query heads) | Size (float16) |
|---|---|---|
| 512 | 16 | ~8 MB |
| 2 048 | 16 | ~128 MB |
| 4 096 | 32 | ~1 GB |

Tensors are gzip-compressed before serialisation, typically achieving 3–5×
reduction. The persistent Q cache is FIFO-capped at 32 768 slots (~128 MB
worst-case) to bound memory growth across requests.

Capture is disabled by default. No automatic hard cap on response size is
enforced — for very long sequences, callers are expected to limit scope via
`attn_capture_layers`.

---

## Implementation

| Component | Change |
|------------|--------|
| `attn_capture.py` (new, ~600 LOC) | Q buffering, Q×Kᵀ computation, GQA/MQA support, shared-memory IPC |
| `gpu_model_runner.py` | Hook into forward pass; call `capture()` post-request |
| `attention.py` | Buffer Q tensor to capture hook |
| `output_processor.py` | Load snapshot from shared memory and attach to `RequestOutput` |
| `protocol.py` | Add new request/response fields |
| `arg_utils.py`, `cache.py` | Add CLI flags and configuration |

`attn_capture.py` core logic is ~350 lines; the remainder is IPC utilities
and validation helpers. Can be split if preferred.

---

## Example Usage

```python
from openai import OpenAI
import json, base64, gzip, numpy as np

client = OpenAI(base_url="http://localhost:8000/v1", api_key="EMPTY")

raw = client.chat.completions.with_raw_response.create(
    model="Qwen/Qwen2.5-VL-3B-Instruct",
    messages=[{"role": "user", "content": "Hello!"}],
    extra_body={"attn_capture": 1, "attn_capture_layers": "8,16"},
)

resp = json.loads(raw.content)

for item in resp["attn_capture_data"]:
    blob = gzip.decompress(base64.b64decode(item["data"]))
    scores = np.frombuffer(blob, dtype=item["dtype"]).reshape(item["shape"])
    print(f"Layer {item['layer_idx']}: shape={scores.shape}")
    # scores: [T, H, T] — rows=query positions, cols=key positions
```

See `examples/offline_inference/attention_instrumentation/` for full examples
including multimodal token classification and cross-modal attention analysis.

---

## Test Results

Tested on **Qwen2.5-VL-3B-Instruct** and **Gemma-3-4B-IT** across 9 input
groups × 6 capture configurations = **54 cases per model, 108 total**.
All 108/108 cases passed.

Test script: [Parkprogrammer/vllm_capture](https://github.com/Parkprogrammer/vllm_capture)

Summary:
- No regression when `attn_capture=0` (instrumentation-enabled server, capture disabled per-request)
- Overhead is a fixed per-request cost; does not scale with sequence length
- Medium/long prompt groups show near-parity with baseline throughput

<details>
<summary>Full per-group breakdown</summary>

Input groups: `text·{short,medium,long}`, `text+image·{short,medium,long}`, `image only·{short,medium,long}`

Capture configs: `off` (attn_capture=0), `early` (layer 2), `mid` (layer 8), `late` (layer 15), `multi` (2,8,15), `all`

### Qwen2.5-VL-3B-Instruct

| Group | Pass | AvgLat | AvgTok/s | Off Tok/s | BaseTok/s | AvgPTok |
|---|---|---|---|---|---|---|
| text · short | 6/6 | 0.69s | 34.9 | 35.9 | 48.0 | 28 |
| text · medium | 6/6 | 1.16s | 41.4 | 51.2 | 50.8 | 76 |
| text · long | 6/6 | 1.19s | 40.4 | 50.3 | 50.0 | 432 |
| text+image · short | 6/6 | 0.94s | 25.6 | 10.8 | 47.1 | 494 |
| text+image · medium | 6/6 | 1.21s | 39.7 | 48.6 | 49.6 | 542 |
| text+image · long | 6/6 | 1.18s | 40.6 | 50.3 | 49.6 | 898 |
| image only · short | 6/6 | 0.81s | 39.7 | 50.1 | 45.4 | 485 |
| image only · medium | 6/6 | 1.18s | 40.8 | 50.1 | 45.7 | 485 |
| image only · long | 6/6 | 1.19s | 40.3 | 48.5 | 47.3 | 485 |

### Gemma-3-4B-IT

| Group | Pass | AvgLat | AvgTok/s | Off Tok/s | BaseTok/s | AvgPTok |
|---|---|---|---|---|---|---|
| text · short | 6/6 | 0.93s | 23.7 | 27.1 | 26.8 | 18 |
| text · medium | 6/6 | 1.89s | 25.4 | 28.6 | 28.7 | 64 |
| text · long | 6/6 | 1.88s | 25.5 | 28.7 | 28.7 | 423 |
| text+image · short | 6/6 | 1.35s | 16.4 | 17.9 | 17.6 | 278 |
| text+image · medium | 6/6 | 2.73s | 17.6 | 18.8 | 19.0 | 324 |
| text+image · long | 6/6 | 2.77s | 17.4 | 18.7 | 18.8 | 683 |
| image only · short | 6/6 | 1.86s | 17.2 | 18.9 | 18.7 | 268 |
| image only · medium | 6/6 | 2.77s | 17.3 | 18.1 | 19.0 | 268 |
| image only · long | 6/6 | 2.78s | 17.3 | 18.7 | 18.7 | 268 |

Column legend:
- **AvgTok/s**: mean tok/s across all 6 capture configs including `all`
- **Off Tok/s**: `attn_capture=0` on instrumentation-enabled server
- **BaseTok/s**: plain vLLM server with no `--enable-attention-instrumentation` flag (2 warm-up requests; indicative, single GPU)

</details>

### CODE REVIEW

**Review by DarkLight1337 — COMMENTED** [review]
We have previously rejected returning hidden states directly in the request output because it adds too much IPC overhead (see https://github.com/vllm-project/vllm/pull/15434#issuecomment-2816502014), the same could be said for this PR. I suggest following a similar approach as https://github.com/vllm-project/vllm/issues/33118

File: vllm/v1/engine/output_processor.py [hunk 1]
```diff
@@ -360,6 +361,19 @@ def _new_request_output(
...
 
+        # If prompt embeds were used, put placeholder prompt token ids
+        prompt_token_ids = self.prompt_token_ids
+        if prompt_token_ids is None and self.prompt_embeds is not None:
+            prompt_token_ids = [0] * len(self.prompt_embeds)
+
+        # Load attention capture data if capture was requested
+        attn_capture_data = None
+        if finished and getattr(self, "attn_capture_enabled", False):
+            from vllm.model_executor.layers.attention.attn_capture import (
```
[the remark this answers is not in the document]
Comment (Parkprogrammer) [reply vllm/v1/engine/output_processor.py:372]: The blocking `load_attn_snapshot()` call will be removed from output_processor. In the rework, the engine returns only an **out-of-band handle** ([via kv_transfer_params](url)), and no polling happens on the engine loop.

File: vllm/model_executor/layers/attention/attn_capture.py [hunk 2]
```diff
@@ -0,0 +1,604 @@
...
+
+        # Extract slot_ids using attn_metadata from the gpu_model_runner process
+        slot_ids = attn_metadata.slot_mapping
+        if query.shape[0] != slot_ids.shape[0]:
+            return
+
+        # Detaching the query tensor for buffering. Severing CUDA computation trace
+        # Query tensor stays in the dict for all requests
+        try:
+            query_cpu = query.detach().cpu().clone()
```
[the remark this answers is not in the document]
Comment (Parkprogrammer) [reply vllm/model_executor/layers/attention/attn_capture.py:115]: The `query.detach().cpu().clone()` path introduces an implicit GPU↔CPU sync. I will replace it with an async **D2H** copy into pinned memory on a dedicated stream to **avoid blocking the forward hot path**.

File: vllm/model_executor/layers/attention/attn_capture.py [hunk 3]
```diff
@@ -0,0 +1,604 @@
...
+        except FileNotFoundError:
+            time.sleep(0.05)
+            continue
+        size = struct.unpack_from("Q", mem.buf, 0)[0]
+        if size == 0:
+            mem.close()
+            time.sleep(0.01)
+            continue
+        try:
+            data = pickle.loads(bytes(mem.buf[8 : 8 + size]))
```
[the remark this answers is not in the document]
Comment (Parkprogrammer) [reply vllm/model_executor/layers/attention/attn_capture.py:592]: I will remove pickle from the IPC path and replace it with a safe binary format (fixed struct header + raw tensor bytes). No object deserialization.

File: vllm/model_executor/layers/attention/attn_capture.py [hunk 4]
```diff
@@ -0,0 +1,604 @@
...
+        k = kv_cache[0, slot_tensor]
+    else:
+        raise ValueError(
+            f"Unsupported KV cache layout: ndim={kv_cache.ndim} shape={list(shape)}"
+        )
+
+    return k.to(dtype)  # NOTE(jehyun): Uniform dtype for downstream float16 computation
+
+
+def compute_qk_attention(
```
[the remark this answers is not in the document]
Comment (Parkprogrammer) [reply vllm/model_executor/layers/attention/attn_capture.py:410]: will add hard per-request resource caps (e.g., `--max-attn-capture-tokens`, `--max-attn-capture-bytes`, and/or max layers). Requests exceeding limits will be rejected with an explicit status instead of attempting allocation.

File: vllm/v1/worker/gpu_model_runner.py [hunk 5]
```diff
@@ -6264,6 +6295,62 @@ def timed_encoder_operation(
...
+
+        assert self.attn_capture is not None
+        self.attn_capture.runtime_enabled_this_step = capture_slots is not None
+        self.attn_capture.capture_slots = capture_slots
+
+    def _capture_finished_request(self, req_id: str) -> None:
+        """Capture attention and clean up buffers for a finished request."""
+        req_state = self.requests.get(req_id)
+        if not req_state:
+            return
```
[the remark this answers is not in the document]
Comment (Parkprogrammer) [reply vllm/v1/worker/gpu_model_runner.py:6335]: `capture()` will be offloaded so the **worker execution loop is non-blocking**. The worker will enqueue a capture job and proceed; the capture task performs **post-forward compute** + **compression** asynchronously and writes the result out-of-band.
